### DISCUSSION TITLE
Receive files from server

### DISCUSSION THREAD

**Discussion by frogcrush** [christianhelle/refitter#507] [Q&A]
I'm trying to transfer files from my ASP.NET Core app to my desktop apps that use Refitter to generate API libraries. 

Using the following Refitter file:
```json
{
    "openApiPath": "blah.json",
    "namespace": "X.Clients",
    "naming": {
        "useOpenApiTitle": false,
        "interfaceName": "XClient"
    },
    "outputFolder": "GeneratedCode",
    "outputFilename": "XAPI.cs",
    "returnIApiResponse": true,
    "generateContracts": false,
    "optionalParameters": false,
    "generateDefaultAdditionalProperties": false,
    "codeGeneratorSettings": {
        "dateType": "System.DateOnly",
        "dateTimeType": "System.DateTimeOffset",
        "timeType": "System.TimeOnly",
        "timeSpanType": "System.TimeSpan"
    }
}
```

I had an issue where it kept generating endpoints with IApiResponse<FIleResponse> but FileResponse did not exist. I turned on GenerateContracts and copied the FileResponse model:

```csharp
    public partial class FileResponse : IDisposable
    {
        private IDisposable _client;
        private IDisposable _response;

        public int StatusCode { get; private set; }

        public IReadOnlyDictionary<string, IEnumerable<string>> Headers { get; private set; }

        public Stream Stream { get; private set; }

        public bool IsPartial
        {
            get { return StatusCode == 206; }
        }

        public FileResponse(int statusCode, IReadOnlyDictionary<string, IEnumerable<string>> headers, Stream stream, IDisposable client, IDisposable response)
        {
            StatusCode = statusCode;
            Headers = headers;
            Stream = stream;
            _client = client;
            _response = response;
        }

        public void Dispose()
        {
            Stream.Dispose();
            _response?.Dispose();

            _client?.Dispose();
        }
    }
```

But when I try and get the file at runtime it hits a serialization error. I'm thinking maybe IApiResult is trying to deserialize the file into the FileResponse?

I tried setting responseTypeOverride:

```json
  "responseTypeOverride": {
    "File_Upload": "IApiResponse",
    "File_Download": "System.Net.Http.HttpContent"
  },
```

But there was no effect.

What is the correct approach to be downloading files?

**Comment by christianhelle** [comment]
@frogcrush Thanks for taking the time to bring this up. I have never tried using Refit to download files directly. I would build something like this on the backend by creating an endpoint with a response payload that includes a short-lived URL and some metadata to the actual file to be downloaded.

**Comment by frogcrush** [comment]
Well it is not downloading static files, they are dynamic from the server (server is returning FileStreamResult) - when using just Refit before I could use HttpResponseMessage instead of IApiResponse and access the stream without issue. Is there any way to have Refitter detect if an endpoint is returning ``application/octet-stream`` and use HttpResponseMessage instead of IApiResult?

I thought that's what responseTypeOverride would do but it doesn't appear to (at least with my configuration).

I can see that Refitter is correctly seeing that the endpoints return ``application/octet-stream``:
```csharp
        [Headers("Accept: application/octet-stream")]
        [Get("/api/projects/{id}/coverphoto")]
        Task<IApiResponse<FileResponse>> GetProjectCoverPhoto(string id);
```

**Reply by christianhelle** [reply]
> Is there any way to have Refitter detect if an endpoint is returning `application/octet-stream` and use HttpResponseMessage instead of IApiResult?

I like this idea of defaulting to `HttpContent` or `HttpResponseMessage` for `application/octet-stream` content types. We currently don't have that feature. 

I'm currently busy with other things and it might take some time before I can build this feature. Will you be interested in implementing this @frogcrush? If you create a pull request I'll make sure the feature gets in

**Comment by christianhelle** [comment]
> I tried setting responseTypeOverride:
> 
> ```json
>   "responseTypeOverride": {
>     "File_Upload": "IApiResponse",
>     "File_Download": "System.Net.Http.HttpContent"
>   },
> ```
> But there was no effect.

@frogcrush the `responsetypeOverride` setting uses the Operation ID as the key, and the type as the value

For example:
```json
 "responseTypeOverride": {
     "getPetById": "IApiResponse<Pet>",
     "getPetImageById": "System.Net.Http.HttpContent"
 }
```
